repo/kripke/package.py

```diff
@@ -0,0 +1,125 @@
+# Copyright 2023 Lawrence Livermore National Security, LLC and other
+# Benchpark Project Developers. See the top-level COPYRIGHT file for details.
+#
+# SPDX-License-Identifier: Apache-2.0
+
+from spack.package import *
+
+
+class Kripke(CMakePackage, CudaPackage, ROCmPackage):
+    """Kripke is a simple, scalable, 3D Sn deterministic particle
+    transport proxy/mini app.
+    """
+
+    homepage = "https://computing.llnl.gov/projects/co-design/kripke"
+    git = "https://github.com/LLNL/Kripke.git"
+
+    tags = ["proxy-app"]
+
+    maintainers("vsrana01")
+
+    license("BSD-3-Clause")
+
+    version("develop", branch="develop", submodules=False)
+    version(
+        "updateRCU2024.02", branch="task/chen59/updateRCU2024.02", submodules=False)
+    version(
+        "1.2.4", submodules=False, tag="v1.2.4", commit="d85c6bc462f17a2382b11ba363059febc487f771"
+    )
+    version(
+        "1.2.3", submodules=True, tag="v1.2.3", commit="66046d8cd51f5bcf8666fd8c810322e253c4ce0e"
+    )
+    version(
+        "1.2.2",
+        submodules=True,
+        tag="v1.2.2-CORAL2",
+        commit="a12bce71e751f8f999009aa2fd0839b908b118a4",
+    )
+    version(
+        "1.2.1",
+        submodules=True,
+        tag="v1.2.1-CORAL2",
+        commit="c36453301ddd684118bb0fb426cfa62764d42398",
+    )
+    version(
+        "1.2.0",
+        submodules=True,
+        tag="v1.2.0-CORAL2",
+        commit="67e4b0a2f092009d61f44b5122111d388a3bec2a",
+    )
+
+    variant("mpi", default=True, description="Build with MPI.")
+    variant("openmp", default=False, description="Build with OpenMP enabled.")
+    variant("caliper", default=False, description="Build with Caliper support enabled.")
+
+    depends_on("mpi", when="+mpi")
+    depends_on("caliper", when="+caliper")
+    depends_on("adiak@0.4:", when="+caliper")
+    
+    aligned_versions = ["2024.02"]
+    constrain_chai = list(f"^chai@{v}" for v in aligned_versions)
+    requires(*aligned_versions, policy="one_of", msg="Pick a chai version with proper alignment w/RAJA & Umpire")
+    
+    for v in aligned_versions:
+      depends_on(f"raja@{v}~exercises~examples", when=f"^chai@{v}")
+      depends_on(f"umpire@{v}~examples", when=f"^chai@{v}")
+      depends_on(f"chai@{v}~examples+raja", when=f"^chai@{v}")
```

Comment:
@scheibelp is this how I would specify chai in here?  I do want to build chai ~examples+raja.

Also, in your example you had constrain chai - but just using raja and umpire in the loop.  Is there a reason one was chosen?

Is the list in constrain chai then getting the list of minor versions as well?  I don't understand how the minor versions would work here.

Reply:
> Also, in your example you had constrain chai - but just using raja and umpire in the loop. Is there a reason one was chosen?

Speaking loosely, these are edge constraints, and two edges covers three nodes

`depends_on(f"raja@{v}~exercises~examples", when=f"^chai@{v}")` for example imposes a version constraint on `raja` based on `chai`.

The third constraint will have no effect (it won't cause a problem, but won't change how anything concretizes).

Note that if you don't force `chai` to be one of the versions in aligned versions, Spack will feel free to pick other versions of `chai` that do not have version-locking constraints you set up here, so

```
constrain_chai = list(f"^chai@{v}" for v in aligned_versions)
requires(*constrain_chai, policy="one_of", msg="Pick a chai version with proper alignment w/RAJA & Umpire")
```

(I'll note there was a typo where I mentioned this to you originally, and the example directly above fixes that)